File: example/Lab05_1.py
from Crypto.Cipher import DES
from Crypto.Util.Padding import pad, unpad
import time
import os
import sys

class DESFileEncryption:
    def __init__(self, key):
        if len(key) != 8:
            raise ValueError("DES key phải có độ dài 8 bytes")
        self.key = key.encode() if isinstance(key, str) else key
    
    def encrypt_file(self, input_file, output_file, mode='ECB', padding='PKCS5'):
        start_time = time.time()
        
        with open(input_file, 'rb') as f:
            plaintext = f.read()
        
        if mode == 'ECB':
            cipher = DES.new(self.key, DES.MODE_ECB)
            iv = None
        elif mode == 'CBC':
            cipher = DES.new(self.key, DES.MODE_CBC)
            iv = cipher.iv
        else:
            raise ValueError("Mode phải là ECB hoặc CBC")
       
        if padding == 'PKCS5':
            plaintext = pad(plaintext, DES.block_size)
        elif padding == 'NoPadding':
            
            if len(plaintext) % DES.block_size != 0:
                raise ValueError("Với NoPadding, dữ liệu phải là bội số của 8 bytes")
        else:
            raise ValueError("Padding phải là PKCS5 hoặc NoPadding")
        
       
        ciphertext = cipher.encrypt(plaintext)
        
        with open(output_file, 'wb') as f:
            if mode == 'CBC':
                f.write(iv)
            f.write(ciphertext)
        
        end_time = time.time()
        elapsed = end_time - start_time
        
        file_size = os.path.getsize(input_file)
        print(f"Mã hóa thành công!")
        print(f"Mode: DES/{mode}/{padding}")
        print(f"Kích thước file: {file_size / (1024*1024):.2f} MB")
        print(f"Thời gian mã hóa: {elapsed:.4f} giây")
        print(f"Tốc độ: {(file_size / (1024*1024)) / elapsed:.2f} MB/s")
        
        return elapsed
    
    def decrypt_file(self, input_file, output_file, mode='ECB', padding='PKCS5'):
        start_time = time.time()
        
        with open(input_file, 'rb') as f:
            data = f.read()
        
        if mode == 'ECB':
            cipher = DES.new(self.key, DES.MODE_ECB)
            ciphertext = data
        elif mode == 'CBC':
            iv = data[:8]
            ciphertext = data[8:]
            cipher = DES.new(self.key, DES.MODE_CBC, iv=iv)
        else:
            raise ValueError("Mode phải là ECB hoặc CBC")
        
        plaintext = cipher.decrypt(ciphertext)
        
        if padding == 'PKCS5':
            plaintext = unpad(plaintext, DES.block_size)
        
        with open(output_file, 'wb') as f:
            f.write(plaintext)
        
        end_time = time.time()
        elapsed = end_time - start_time
        
        file_size = os.path.getsize(output_file)
        print(f"Giải mã thành công!")
        print(f"Mode: DES/{mode}/{padding}")
        print(f"Kích thước file: {file_size / (1024*1024):.2f} MB")
        print(f"Thời gian giải mã: {elapsed:.4f} giây")
        print(f"Tốc độ: {(file_size / (1024*1024)) / elapsed:.2f} MB/s")
        
        return elapsed


def main():
    print("=" * 60)
    print("CHƯƠNG TRÌNH MÃ HÓA/GIẢI MÃ TẬP TIN VỚI DES")
    print("=" * 60)
    
    action = input("\nChọn hành động (1-Mã hóa, 2-Giải mã): ").strip()
    
    if action not in ['1', '2']:
        print("Lựa chọn không hợp lệ!")
        return
    
    input_file = input("Nhập tên file đầu vào: ").strip()
    
    if not os.path.exists(input_file):
        print(f"File {input_file} không tồn tại!")
        return
    
    key_file = input("Nhập tên file chứa khóa (hoặc Enter để nhập khóa trực tiếp): ").strip()
    
    if key_file and os.path.exists(key_file):
        with open(key_file, 'rb') as f:
            key = f.read()[:8]
    else:
        key_input = input("Nhập khóa (8 ký tự): ").strip()
        if len(key_input) != 8:
            print("Khóa phải có độ dài 8 ký tự!")
            return
        key = key_input
    
    print("\nChế độ mã hóa:")
    print("1. DES/ECB/PKCS5Padding")
    print("2. DES/ECB/NoPadding")
    print("3. DES/CBC/PKCS5Padding")
    print("4. DES/CBC/NoPadding")
    mode_choice = input("Chọn chế độ (1-4): ").strip()
    
    mode_map = {
        '1': ('ECB', 'PKCS5'),
        '2': ('ECB', 'NoPadding'),
        '3': ('CBC', 'PKCS5'),
        '4': ('CBC', 'NoPadding')
    }
    
    if mode_choice not in mode_map:
        print("Lựa chọn không hợp lệ!")
        return
    
    mode, padding = mode_map[mode_choice]
    
    des = DESFileEncryption(key)
    
    try:
        if action == '1':
            # Encode
            output_file = "output.enc"
            des.encrypt_file(input_file, output_file, mode, padding)
            print(f"\nFile đã được mã hóa và lưu vào: {output_file}")
        else:
            # Decode
            output_file = "output.dec"
            des.decrypt_file(input_file, output_file, mode, padding)
            print(f"\nFile đã được giải mã và lưu vào: {output_file}")
    
    except Exception as e:
        print(f"\nLỗi: {str(e)}")
        import traceback
        traceback.print_exc()


def test_performance():
    print("\n" + "=" * 60)
    print("TEST HIỆU NĂNG VỚI CÁC FILE KHÁC NHAU")
    print("=" * 60)
    
    key = "abcdEFGH"
    des = DESFileEncryption(key)
    
    test_files = {
        "10MB": "test_10mb.bin",
        "100MB": "test_100mb.bin",
        "1GB": "test_1gb.bin"
    }
    
    modes = [
        ('ECB', 'PKCS5'),
        ('ECB', 'NoPadding'),
        ('CBC', 'PKCS5'),
        ('CBC', 'NoPadding')
    ]
    
    for size_name, filename in test_files.items():
        if not os.path.exists(filename):
            print(f"\nTạo file test {size_name}...")
            size_bytes = {'10MB': 10*1024*1024, '100MB': 100*1024*1024, '1GB': 1024*1024*1024}
            with open(filename, 'wb') as f:
                chunk_size = 1024 * 1024  # 1MB
                for _ in range(size_bytes[size_name] // chunk_size):
                    f.write(os.urandom(chunk_size))
    
    results = []
    for size_name, filename in test_files.items():
        if not os.path.exists(filename):
            print(f"\nBỏ qua {size_name}: File không tồn tại")
            continue
        
        print(f"\n{'=' * 60}")
        print(f"Test với file {size_name}")
        print(f"{'=' * 60}")
        
        for mode, padding in modes:
            print(f"\n--- Mode: DES/{mode}/{padding} ---")
            try:
                enc_time = des.encrypt_file(filename, "temp.enc", mode, padding)
                dec_time = des.decrypt_file("temp.enc", "temp.dec", mode, padding)
                
                results.append({
                    'size': size_name,
                    'mode': f"{mode}/{padding}",
                    'enc_time': enc_time,
                    'dec_time': dec_time
                })
                
                os.remove("temp.enc")
                os.remove("temp.dec")
            except Exception as e:
                print(f"Lỗi: {str(e)}")
    
    print("\n" + "=" * 60)
    print("TỔNG KẾT KẾT QUẢ")
    print("=" * 60)
    for r in results:
        print(f"\n{r['size']} - {r['mode']}:")
        print(f"  Mã hóa: {r['enc_time']:.4f}s")
        print(f"  Giải mã: {r['dec_time']:.4f}s")


if __name__ == "__main__":
    import sys
    
    if len(sys.argv) > 1 and sys.argv[1] == '--test':
        test_performance()
    else:
        main()

File: code/Lab05_1.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.*;
import javax.crypto.*;
import javax.crypto.spec.*;
import java.util.Scanner;

/**
 * Lab05_1: DES File Encryption/Decryption
 * 
 * WARNING: This implementation uses DES and ECB mode for EDUCATIONAL PURPOSES ONLY.
 * 
 * Security issues:
 * - DES has 56-bit key size (broken by modern standards)
 * - ECB mode does not hide data patterns (cryptographically weak)
 * 
 * For production use:
 * - Use AES-256 instead of DES
 * - Use CBC or GCM mode instead of ECB
 * 
 * Implements 4 encryption modes:
 * 1. DES/ECB/PKCS5Padding
 * 2. DES/ECB/NoPadding
 * 3. DES/CBC/PKCS5Padding
 * 4. DES/CBC/NoPadding
 */
public class Lab05_1 {
    
    /**
     * Encrypts a file using DES with specified mode and padding
     */
    public static double encryptFile(String inputFile, String outputFile, 
                                     byte[] key, String mode, String padding) 
            throws Exception {
        
        long startTime = System.nanoTime();
        
        // Validate key length
        if (key.length != 8) {
            throw new IllegalArgumentException("DES key must be 8 bytes");
        }
        
        // Create secret key
        SecretKeySpec secretKey = new SecretKeySpec(key, "DES");
        
        // Create cipher
        String transformation = "DES/" + mode + "/" + padding;
        Cipher cipher = Cipher.getInstance(transformation);
        
        // Initialize cipher
        if (mode.equals("CBC")) {
            cipher.init(Cipher.ENCRYPT_MODE, secretKey);
        } else {
            cipher.init(Cipher.ENCRYPT_MODE, secretKey);
        }
        
        // Read input file
        byte[] inputBytes = Files.readAllBytes(Paths.get(inputFile));
        
        // Check alignment for NoPadding
        if (padding.equals("NoPadding") && inputBytes.length % 8 != 0) {
            throw new IllegalArgumentException(
                "With NoPadding, data length must be multiple of 8 bytes. " +
                "Current length: " + inputBytes.length);
        }
        
        // Encrypt
        byte[] outputBytes = cipher.doFinal(inputBytes);
        
        // Write output file
        try (FileOutputStream fos = new FileOutputStream(outputFile)) {
            // Write IV for CBC mode
            if (mode.equals("CBC")) {
                byte[] iv = cipher.getIV();
                fos.write(iv);
            }
            fos.write(outputBytes);
        }
        
        long endTime = System.nanoTime();
        double elapsed = (endTime - startTime) / 1_000_000_000.0;
        
        // Print statistics
        File inFile = new File(inputFile);
        double fileSizeMB = inFile.length() / (1024.0 * 1024.0);
        System.out.println("Encryption successful!");
        System.out.println("Mode: " + transformation);
        System.out.println(String.format("File size: %.2f MB", fileSizeMB));
        System.out.println(String.format("Encryption time: %.4f seconds", elapsed));
        if (elapsed > 0) {
            System.out.println(String.format("Speed: %.2f MB/s", fileSizeMB / elapsed));
        }
        
        return elapsed;
    }
    
    /**
     * Decrypts a file using DES with specified mode and padding
     */
    public static double decryptFile(String inputFile, String outputFile, 
                                     byte[] key, String mode, String padding) 
            throws Exception {
        
        long startTime = System.nanoTime();
        
        // Validate key length
        if (key.length != 8) {
            throw new IllegalArgumentException("DES key must be 8 bytes");
        }
        
        // Create secret key
        SecretKeySpec secretKey = new SecretKeySpec(key, "DES");
        
        // Create cipher
        String transformation = "DES/" + mode + "/" + padding;
        Cipher cipher = Cipher.getInstance(transformation);
        
        // Read input file
        byte[] inputBytes = Files.readAllBytes(Paths.get(inputFile));
        
        // Extract IV for CBC mode
        byte[] ciphertext;
        if (mode.equals("CBC")) {
            byte[] iv = new byte[8];
            System.arraycopy(inputBytes, 0, iv, 0, 8);
            ciphertext = new byte[inputBytes.length - 8];
            System.arraycopy(inputBytes, 8, ciphertext, 0, ciphertext.length);
            
            IvParameterSpec ivSpec = new IvParameterSpec(iv);
            cipher.init(Cipher.DECRYPT_MODE, secretKey, ivSpec);
        } else {
            ciphertext = inputBytes;
            cipher.init(Cipher.DECRYPT_MODE, secretKey);
        }
        
        // Decrypt
        byte[] outputBytes = cipher.doFinal(ciphertext);
        
        // Write output file
        Files.write(Paths.get(outputFile), outputBytes);
        
        long endTime = System.nanoTime();
        double elapsed = (endTime - startTime) / 1_000_000_000.0;
        
        // Print statistics
        double fileSizeMB = outputBytes.length / (1024.0 * 1024.0);
        System.out.println("Decryption successful!");
        System.out.println("Mode: " + transformation);
        System.out.println(String.format("File size: %.2f MB", fileSizeMB));
        System.out.println(String.format("Decryption time: %.4f seconds", elapsed));
        if (elapsed > 0) {
            System.out.println(String.format("Speed: %.2f MB/s", fileSizeMB / elapsed));
        }
        
        return elapsed;
    }
    
    /**
     * Interactive main menu
     */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("============================================================");
        System.out.println("DES FILE ENCRYPTION/DECRYPTION PROGRAM");
        System.out.println("============================================================");
        
        try {
            // Choose action
            System.out.print("\nChoose action (1-Encrypt, 2-Decrypt): ");
            String action = scanner.nextLine().trim();
            
            if (!action.equals("1") && !action.equals("2")) {
                System.out.println("Invalid choice!");
                return;
            }
            
            // Input file
            System.out.print("Enter input file name: ");
            String inputFile = scanner.nextLine().trim();
            
            File inFile = new File(inputFile);
            if (!inFile.exists()) {
                System.out.println("File " + inputFile + " does not exist!");
                return;
            }
            
            // Key
            System.out.print("Enter key file name (or press Enter to type key directly): ");
            String keyFile = scanner.nextLine().trim();
            
            byte[] key;
            if (!keyFile.isEmpty() && new File(keyFile).exists()) {
                key = Files.readAllBytes(Paths.get(keyFile));
                if (key.length < 8) {
                    System.out.println("Key file too short! Need 8 bytes.");
                    return;
                }
                // Take first 8 bytes
                byte[] keyBytes = new byte[8];
                System.arraycopy(key, 0, keyBytes, 0, 8);
                key = keyBytes;
            } else {
                System.out.print("Enter key (8 characters): ");
                String keyStr = scanner.nextLine().trim();
                if (keyStr.length() != 8) {
                    System.out.println("Key must be 8 characters!");
                    return;
                }
                key = keyStr.getBytes(StandardCharsets.UTF_8);
            }
            
            // Mode selection
            System.out.println("\nEncryption mode:");
            System.out.println("1. DES/ECB/PKCS5Padding");
            System.out.println("2. DES/ECB/NoPadding");
            System.out.println("3. DES/CBC/PKCS5Padding");
            System.out.println("4. DES/CBC/NoPadding");
            System.out.print("Choose mode (1-4): ");
            String modeChoice = scanner.nextLine().trim();
            
            String mode, padding;
            switch (modeChoice) {
                case "1":
                    mode = "ECB";
                    padding = "PKCS5Padding";
                    break;
                case "2":
                    mode = "ECB";
                    padding = "NoPadding";
                    break;
                case "3":
                    mode = "CBC";
                    padding = "PKCS5Padding";
                    break;
                case "4":
                    mode = "CBC";
                    padding = "NoPadding";
                    break;
                default:
                    System.out.println("Invalid choice!");
                    return;
            }
            
            // Perform encryption/decryption
            String outputFile;
            if (action.equals("1")) {
                outputFile = "output.enc";
                encryptFile(inputFile, outputFile, key, mode, padding);
                System.out.println("\nFile encrypted and saved to: " + outputFile);
            } else {
                outputFile = "output.dec";
                decryptFile(inputFile, outputFile, key, mode, padding);
                System.out.println("\nFile decrypted and saved to: " + outputFile);
            }
            
        } catch (Exception e) {
            System.out.println("\nError: " + e.getMessage());
            e.printStackTrace();
        } finally {
            scanner.close();
        }
    }
}
